ui/frontend/reducers/code.ts

```diff
@@ -19,8 +19,30 @@ export default function code(state = DEFAULT, action: Action): State {
     case ActionType.AddMainFunction:
       return `${state}\n\n${DEFAULT}`;
 
-    case ActionType.AddImport:
-      return action.code + state;
+    case ActionType.ApplySuggestion:
+      const state_lines = state.split('\n');
+      const startline = action.startline - 1;
+      const endline = action.endline - 1;
+      const startcol = action.startcol - 1;
+      const endcol = action.endcol - 1;
+      if (startline == endline) {
+        state_lines[startline] = state_lines[startline].substring(0, startcol) +
+          state_lines[startline].substring(endcol);
+      } else {
+        if (state_lines.length > startline) {
+          state_lines[startline] = state_lines[startline].substring(0, startcol);
+        }
+        if (state_lines.length > endline) {
+          state_lines[endline] = state_lines[endline].substring(endcol);
+        }
+        if (endline - startline > 1) {
+          state_lines.splice(startline + 1, endline - startline - 1);
+        }
+      }
+      state_lines[startline] = state_lines[startline].substring(0, startcol) +
+        action.suggestion + state_lines[startline].substring(startcol);
+      state = state_lines.join('\n');
+      return state;
```

Comment:
This logic is tricky enough that I can't just glance at it and feel very confident about what it's doing. It'd be helpful to extract out the core logic here and write some unit tests to exercise the various `if`s. Something like this sketch:

```ts
    case ActionType.ApplySuggestion:
       const startline = action.startline - 1;
       const endline = action.endline - 1;
       const startcol = action.startcol - 1;
       const endcol = action.endcol - 1;
       return someLogic(state);
```